pints/_diagnostics.py

```diff
@@ -55,72 +55,185 @@ def effective_sample_size(samples):
+    # Get number of samples
+    n = chains.shape[1]
 
+    # Compute within-chain mean
+    within_chain_means = np.mean(chains, axis=1)
 
-def rhat_all_params(chains):
+    # Compute variance across chains of within-chain means
+    between_chain_var = np.var(within_chain_means, axis=0, ddof=1)
+
+    # Weight variance with number of samples per chain
+    b = n * between_chain_var
+
+    return b
+
+
+def rhat(chains, warm_up=0.0):
     r"""
-    Calculates :math:`\hat{R} = \sqrt{(W(n - 1) / n + (1 / n) B) / W}` as per
-    [1]_ for all parameters. It does this after splitting each chain into two.
+    Returns the convergence measure :math:`\hat{R}` for the approximate
+    posterior according to [1]_.
+
+    :math:`\hat{R}` diagnoses convergence by checking mixing and stationarity
+    of :math:`m` chains (at least two, :math:`m\geq 2`). To diminish the
+    influence of starting values, the first portion of each chain can be
+    excluded from the computation. Subsequently, the truncated
+    chains are split in half, resulting in a total number of :math:`m'=2m`
+    chains of length :math:`n'=(1-\text{warm_up})n/2`. The mean of the
+    variances within and between the resulting chains are computed, :math:`W`
+    and :math:`B` respectively. Based on those variances an estimator of the
+    marginal posterior variance is constructed
+
+    .. math::
+        \widehat{\text{var}}^+ = \frac{n'-1}{n'}W + \frac{1}{n'}B,
+
+    The estimator overestimates the variance of the marginal posterior
+    if the chains are not well mixed and stationary, but is unbiased if the
+    original chains equal the target distribution. At the same time, the mean
+    within variance :math:`W` underestimates the marginal posterior variance
+    for finite :math:`n`, but converges to the true variance for
+    :math:`n\rightarrow \infty`. By comparing :math:`\widehat{\text{var}}^+`
+    and :math:`W` the mixing and stationarity of the chains can be quantified
+
+    .. math::
+        \hat{R} = \sqrt{\frac{\widehat{\text{var}}^+}{W}}.
+
+    For well mixed and stationary chains :math:`\hat{R}` will be close to one.
+
+    The mean within :math:`W` and mean between :math:`B` variance of the
+    :math:`m'=2m` chains of length :math:`n'=(1-\text{warm_up})n/2` are defined
+    as
+
+    .. math::
+        W = \frac{1}{m'}\sum _{j=1}^{m'}s_j^2\quad \text{where}\quad
+        s_j^2=\frac{1}{n'-1}\sum _{i=1}^{n'}(\psi _{ij} - \bar{\psi} _j)^2,
+
+    .. math::
+        B = \frac{n'}{m'-1}\sum _{j=1}^{m'}(\bar{\psi} _j - \bar{\psi})^2.
+
+    Here, :math:`\psi _{ij}` is the jth sample of the ith
+    chain, :math:`\bar{\psi _j}=\sum _{i=1}^{n'}\psi _{ij}/n'` is the within
+    chain mean of the parameter :math:`\psi` and
+    :math:`\bar{\psi } = \sum _{j=1}^{m'}\bar{\psi} _{j}/m'` is the between
+    chain mean of the within chain means.
 
     References
     ----------
     ..  [1] "Bayesian data analysis", 3rd edition, Gelman et al., 2014.
```

Comment:
Can we point to the relevant chapter in this? Think it's chapter 11. Also, I think we should cite Stan here, since the splitting of chains in two is what they did (I don't think it's mentioned in Gelman).

Reply:
Yes chapter 11, page 285 equation 11.4. I implemented the method strictly according to this definition. Do you maybe mean the rank-based computation that Stan does, and is not mentioned in this book?

Reply:
Nah, I was just meaning about splitting the chains in two. Does Gelman suggest this? If so, fine to keep as is. I think we can save doing any rank-based stuff for now.